rules/Bundle230/Rector/Class_/AddAnnotationToRepositoryRector.php

```diff
@@ -0,0 +1,160 @@
+    {
+        return new RuleDefinition('Add @extends ServiceEntityRepository<T> annotation to repository classes', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+final class SomeRepository extends ServiceEntityRepository
+{
+    public function __construct(ManagerRegistry $registry)
+    {
+        parent::__construct($registry, SomeEntity::class);
+    }
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+/** @extends ServiceEntityRepository<SomeEntity> */
+final class SomeRepository extends ServiceEntityRepository
+{
+    public function __construct(ManagerRegistry $registry)
+    {
+        parent::__construct($registry, SomeEntity::class);
+    }
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $this->isRepositoryClass($node)) {
+            return null;
+        }
+
+        $entityClass = $this->getEntityClassFromConstructor($node);
+        if ($entityClass === null || $this->hasExtendsAnnotation($node)) {
+            return null;
+        }
+        $this->addAnnotationToNode($node, $entityClass);
+
+        return $node;
+    }
+
+    private function isRepositoryClass(Class_ $class): bool
+    {
+        if ($class->extends instanceof Node\Name) {
+            return $this->getName(
+                $class->extends
+            ) === 'Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository';
+        }
+
+        return false;
+    }
+
+    private function getEntityClassFromConstructor(Class_ $class): ?string
+    {
+        $method = $class->getMethod(MethodName::CONSTRUCT);
+        if ($method === null || $method->stmts === null) {
+            return null;
+        }
+
+        foreach ($method->stmts as $stmt) {
+            if (! $stmt instanceof Node\Stmt\Expression) {
+                continue;
+            }
+
+            $expr = $stmt->expr;
+            if (! $expr instanceof Node\Expr\StaticCall) {
+                continue;
+            }
+
+            if (! $this->isParentConstructorCall($expr)) {
+                continue;
+            }
+
+            $entityClassNode = $expr->args[1]->value ?? null;
+            if (! $entityClassNode instanceof ClassConstFetch) {
+                continue;
+            }
+
+            $entityClass = $entityClassNode->class;
+
+            return $entityClass instanceof Node\Name
+                ? $entityClass->toString()
+                : null;
+        }
+
+        return null;
+    }
+
+    private function addAnnotationToNode(Class_ $class, string $entityClass): void
+    {
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($class);
+        $annotation = sprintf('ServiceEntityRepository<%s>', $entityClass);
```

Comment:
use fully qualified to avoid no use statement

```suggestion
        $annotation = sprintf('\Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository<\%s>', $entityClass);
```

Reply:
Added and updated test case

Reply:
I also updated class names for the test fixtures as i saw you [review comment on another pull request](https://github.com/rectorphp/rector-doctrine/pull/372#discussion_r1992140127)